File: frontend/src/api/recalls.ts
import { apiClient } from './client'

export type RecallSort = 'score' | 'latest'
export type RecallSourceKind =
  | 'structured_api'
  | 'public_notice'
  | 'normalized_public_notice'

export type AuditSummary = {
  audit_id: string
  source_id: string
  module: string
  upstream_status: string
  record_count: number
  transform_version: string
}

export type RecallResult = {
  recall_number: string | null
  product_description: string | null
  reason_for_recall: string | null
  classification: string | null
  status: string | null
  recall_initiation_date: string | null
  distribution_pattern: string | null
  recalling_firm: string | null
  source_type?: string
  source_kind?: RecallSourceKind
  source_record_type?: string
  title?: string | null
  product_name?: string | null
  brand_name?: string | null
  company_name?: string | null
  remedy?: string | null
  record_url?: string | null
  extraction_confidence?: string | null
  source_text_excerpt?: string | null
  risk_score: {
    score: number
    label: string
    components: {
      classification_score: number
      status_score: number
      recency_score: number
      scope_score: number
    }
    score_version: string
  }
  source: {
    name: string
    endpoint: string
    retrieval_timestamp: string
    source_kind?: RecallSourceKind
    source_type?: string
  }
}

export type RecallCheckedSource = {
  source_id: string
  source_name: string
  source_type: string
  endpoint: string
  source_kind: RecallSourceKind
  record_type: string
  upstream_status: string
  record_count: number
  error?: string | null
}

export type RecallSearchResponse = {
  query: string
  raw_query?: string
  normalized_query?: string
  correction_applied?: boolean
  suggestion_message?: string | null
  count: number
  limit: number
  source_name: string
  endpoint: string
  retrieval_timestamp: string
  score_version: string
  sort?: RecallSort
  medical_disclaimer: string
  sources_checked?: RecallCheckedSource[]
  audit: AuditSummary
  results: RecallResult[]
}

export async function searchRecalls(query: string, limit = 5, sort: RecallSort = 'score') {
  const response = await apiClient.get<RecallSearchResponse>(
    `/api/v1/recalls/search`,
    {
      params: {
        q: query,
        limit,
        sort,
      },
    }
  )

  return response.data
}

File: backend/app/schemas/recalls.py
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.audit import AuditSummary


RecallSourceKind = Literal[
    "structured_api",
    "public_notice",
    "normalized_public_notice",
]


class RecallScoreComponents(BaseModel):
    classification_score: int
    status_score: int
    recency_score: int
    scope_score: int


class RecallRiskScore(BaseModel):
    score: int
    label: str
    components: RecallScoreComponents
    score_version: str


class RecallSource(BaseModel):
    name: str
    endpoint: str
    retrieval_timestamp: str
    source_kind: RecallSourceKind = "structured_api"
    source_type: str = "official API record"


class RecallCheckedSource(BaseModel):
    source_id: str
    source_name: str
    source_type: str
    endpoint: str
    source_kind: RecallSourceKind
    record_type: str
    upstream_status: str
    record_count: int
    error: str | None = None


class RecallResult(BaseModel):
    recall_number: str | None
    product_description: str | None
    reason_for_recall: str | None
    classification: str | None
    status: str | None
    recall_initiation_date: str | None
    distribution_pattern: str | None
    recalling_firm: str | None
    source_type: str = "OPENFDA_DRUG_ENFORCEMENT"
    source_kind: RecallSourceKind = "structured_api"
    source_record_type: str = "official API record"
    title: str | None = None
    product_name: str | None = None
    brand_name: str | None = None
    company_name: str | None = None
    remedy: str | None = None
    record_url: str | None = None
    extraction_confidence: str | None = None
    source_text_excerpt: str | None = None
    risk_score: RecallRiskScore
    source: RecallSource


class RecallSemanticMatch(BaseModel):
    record_id: str
    text: str
    similarity_score: float
    explanation: str
    source_name: str | None = None


class RecallSemanticPreview(BaseModel):
    query_text: str
    matches: list[RecallSemanticMatch]
    limitations: list[str]
    preview_version: str
    is_production_ml: bool


class RecallSearchResponse(BaseModel):
    query: str
    raw_query: str
    normalized_query: str
    correction_applied: bool
    suggestion_message: str | None = None
    count: int
    limit: int
    source_name: str
    endpoint: str
    retrieval_timestamp: str
    score_version: str
    medical_disclaimer: str
    sources_checked: list[RecallCheckedSource] = Field(default_factory=list)
    audit: AuditSummary
    semantic_preview: RecallSemanticPreview | None = None
    results: list[RecallResult]
